internal/jobtracker/jobtracker.go

```diff
@@ -0,0 +1,134 @@
+// It periodically performs tasks based on the specified stopInterval duration.
+type Tracker struct {
+	store        Store
+	stopInterval time.Duration
+}
+
+// NewJobTracker creates and returns a new Tracker instance using the provided JobTrackerStore and stopInterval.
+// It returns an error if the store is nil or if stopInterval is not greater than zero.
+func NewJobTracker(store Store, stopInterval time.Duration) (*Tracker, error) {
+	if store == nil {
+		return nil, errors.New("store cannot be nil")
+	}
+	if stopInterval <= 0 {
+		return nil, errors.New("stopInterval must be greater than zero")
+	}
+
+	return &Tracker{
+		stopInterval: stopInterval,
+		store:        store,
+	}, nil
+}
+
+// Run runs a new job and returns the job ID.
+func (j *Tracker) Run(ctx context.Context, jobType string, job func(ctx context.Context) error, maxDuration time.Duration) (uuid.UUID, error) {
+	jobId, err := j.store.AddJob(ctx, jobType)
+	if err != nil {
+		return jobId, err
+	}
+
+	go j.handleJob(ctx, jobId, maxDuration, job)
+
+	return jobId, nil
+}
+
+// handleJob runs a job with a given context, job ID, and deadline.
+// It manages the job's lifecycle, including setting its status in the store, handling retries on store operations,
+// and responding to stop signals or context cancellations. The job is run in a separate goroutine, and its status
+// is updated as running, successful, or failed based on its result or context expiration.
+// If a stop signal is received or the context is canceled or times out, the job is marked as failed.
+// Store operations are retried up to 5 times with a 30-second delay between attempts in case of transient errors.
+func (j *Tracker) handleJob(
+	ctx context.Context,
+	id uuid.UUID,
+	maxDuration time.Duration,
+	job func(ctx context.Context) error,
+) {
+	jobCtx, cancelJob := context.WithTimeout(ctx, maxDuration)
+	defer cancelJob()
+
+	jobErrCh := make(chan error)
+
+	go j.runJob(ctx, jobCtx, id, jobErrCh, job)
+	j.monitorJob(ctx, jobCtx, id, jobErrCh, cancelJob)
+}
+
+func (j *Tracker) runJob(ctx, jobCtx context.Context, id uuid.UUID, jobErrCh chan error, job func(context.Context) error) {
+	if err := j.store.SetJobRunning(ctx, id); err != nil {
+		jobErrCh <- fmt.Errorf("failed to set job running, job not starting: %w", err)
+		return
+	}
+	jobErrCh <- job(jobCtx)
+}
+
+func (j *Tracker) monitorJob(ctx, jobCtx context.Context, id uuid.UUID, jobErrCh chan error, cancelJob context.CancelFunc) {
+	ticker := time.NewTicker(j.stopInterval)
+	defer ticker.Stop()
+
+	// TODO(ale8k): Add monitoring for failed status settings.
+	for {
+		select {
+		case <-jobCtx.Done():
+			err := jobCtx.Err()
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				if err := j.store.SetJobFailed(ctx, id, err); err != nil {
+					zapctx.Error(ctx, "error marking the job as failed", zap.Error(err), zap.String("id", id.String()))
+				}
```

Comment:
If the context was done but it wasn't a cancel or deadlineExceeded, we haven't accounted for timeout errors for example, the job wouldn't have a status?
I thin we should always set the job to failed if the context is done.

Reply:
If those are the only reasons then why are we checking for them? Can't we just set the job to failed anyway?